File: yangxj96-serve/yangxj96-serve-platform/src/main/java/com/yangxj96/saas/server/platform/stream/StreamConsumer.java
package com.yangxj96.saas.server.platform.stream;

import com.alibaba.nacos.shaded.com.google.protobuf.Any;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.rabbitmq.client.Channel;
import com.yangxj96.saas.stream.StreamModel;
import com.yangxj96.saas.stream.tenant.TenantNew;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.support.AmqpHeaders;
import org.springframework.context.annotation.Bean;
import org.springframework.messaging.Message;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.function.Consumer;

/**
 * 消费者
 *
 * @author yangxj96
 * @version 1.0.0
 * @date 2023/4/20 14:24
 */
@Slf4j
@Component
public class StreamConsumer {

    @Resource
    private ObjectMapper om;


    /**
     * 租户消息
     */
    @Bean
    Consumer<Message<StreamModel>> saasTenant() {
        return it -> {
            try {
                var channel = it.getHeaders().get(AmqpHeaders.CHANNEL, Channel.class);
                var delivery = it.getHeaders().get(AmqpHeaders.DELIVERY_TAG, java.lang.Long.class);
                var payload = it.getPayload();
                log.atDebug().log(payload.toString());
                // 新租户创建
                if ("new".equals(payload.getType())) {
                    var obj = om.convertValue(payload.getMessage(), TenantNew.class);
                    log.atDebug().log("新租户:{}", obj);
                }
                if (channel != null && delivery != null) {
                    channel.basicAck(delivery, true);
                } else {
                    log.error("通道/delivery为null");
                }
            } catch (IOException e) {
                log.atError().log("ack操作异常", e);
            }
        };
    }


    /**
     * 平台通用消息
     */
    @Bean
    Consumer<Message<Any>> saasCommon() {
        return it -> {
            try {
                var channel = it.getHeaders().get(AmqpHeaders.CHANNEL, Channel.class);
                var delivery = it.getHeaders().get(AmqpHeaders.DELIVERY_TAG, Long.class);
                var payload = it.getPayload();
                log.atDebug().log(payload.toString());
                if (channel != null && delivery != null) {
                    channel.basicAck(delivery, true);
                } else {
                    log.atError().log("通道/delivery为null");
                }
            } catch (IOException e) {
                log.atError().log("ack操作异常", e);
            }
        };
    }
}

File: yangxj96-serve/yangxj96-serve-platform/src/main/java/com/yangxj96/saas/server/platform/stream/StreamConsumer.kt
package com.yangxj96.saas.server.platform.stream

import com.fasterxml.jackson.databind.ObjectMapper
import com.rabbitmq.client.Channel
import com.yangxj96.saas.stream.StreamModel
import com.yangxj96.saas.stream.tenant.TenantNew
import jakarta.annotation.Resource
import org.slf4j.LoggerFactory
import org.springframework.amqp.support.AmqpHeaders
import org.springframework.context.annotation.Bean
import org.springframework.messaging.Message
import org.springframework.stereotype.Component
import java.io.IOException
import java.util.function.Consumer

/**
 * 消费者
 *
 * @author yangxj96
 * @version 1.0.0
 * @date 2023/4/20 14:24
 */
@Component
class StreamConsumer {

    @Resource
    private lateinit var om: ObjectMapper

    companion object {
        private var log = LoggerFactory.getLogger(this::class.java)
    }

    /**
     * 租户消息
     */
    @Bean
    fun saasTenant(): Consumer<Message<StreamModel>> {
        return Consumer<Message<StreamModel>> {
            try {
                val channel = it.headers[AmqpHeaders.CHANNEL, Channel::class.java]
                val delivery = it.headers[AmqpHeaders.DELIVERY_TAG, java.lang.Long::class.java]
                val payload = it.payload
                // 新租户创建
                if (payload.type == "new") {
                    val obj = om.convertValue(payload.message, TenantNew::class.java)
                    log.info("新租户:{}", obj)
                }
                log.info(payload.toString())
                if (channel != null && delivery != null) {
                    channel.basicAck(delivery.toLong(), true)
                } else {
                    log.error("通道/delivery为null")
                }
            } catch (e: IOException) {
                log.info("ack操作异常")
            }
        }
    }


    /**
     * 平台通用消息
     */
    @Bean
    fun saasCommon(): Consumer<Message<Any>> {
        return Consumer<Message<Any>> {
            try {
                val channel = it.headers[AmqpHeaders.CHANNEL, Channel::class.java]
                val delivery = it.headers[AmqpHeaders.DELIVERY_TAG, java.lang.Long::class.java]
                val payload = it.payload
                log.info(payload.toString())
                if (channel != null && delivery != null) {
                    channel.basicAck(delivery.toLong(), true)
                } else {
                    log.error("通道/delivery为null")
                }
            } catch (e: IOException) {
                log.info("ack操作异常")
            }
        }
    }
}
